### ISSUE TITLE
Compiler error when using clear and crystal-mysql in same code base

### ISSUE THREAD

**Issue by straight-shoota** [anykeyh/clear#177]
When `clear` and `crystal-mysql` are included in the same code base, a compiler error is thrown. Even when `clear` is not even used.

A minimal example is this file:

```cr
require "clear"
require "mysql"
```
Error message:
```
Error: no overload matches 'Hash(String, Array(JSON::Any) | Array(PG::BoolArray) | Array(PG::CharArray) | Array(PG::Float32Array) | Array(PG::Float64Array) | Array(PG::Int16Array) | Array(PG::Int32Array) | Array(PG::Int64Array) | Array(PG::NumericArray) | Array(PG::StringArray) | Array(PG::TimeArray) | Bool | Char | Clear::Expression::UnsafeSql | Float32 | Float64 | Hash(String, JSON::Any) | Int16 | Int32 | Int64 | Int8 | JSON::Any | PG::Geo::Box | PG::Geo::Circle | PG::Geo::Line | PG::Geo::LineSegment | PG::Geo::Path | PG::Geo::Point | PG::Geo::Polygon | PG::Numeric | Slice(UInt8) | String | Time | UInt16 | UInt32 | UInt64 | UInt8 | Nil)#[]=' with types String, (Array(PG::BoolArray) | Array(PG::CharArray) | Array(PG::Float32Array) | Array(PG::Float64Array) | Array(PG::Int16Array) | Array(PG::Int32Array) | Array(PG::Int64Array) | Array(PG::NumericArray) | Array(PG::StringArray) | Array(PG::TimeArray) | Bool | Char | Float32 | Float64 | Int16 | Int32 | Int64 | Int8 | JSON::Any | PG::Geo::Box | PG::Geo::Circle | PG::Geo::Line | PG::Geo::LineSegment | PG::Geo::Path | PG::Geo::Point | PG::Geo::Polygon | PG::Numeric | Slice(UInt8) | String | Time | Time::Span | UInt32 | Nil)
Overloads are:
- Hash(K, V)#[]=(key : K, value : V)
Couldn't find overloads for these types:
- Hash(String, Array(JSON::Any) | Array(PG::BoolArray) | Array(PG::CharArray) | Array(PG::Float32Array) | Array(PG::Float64Array) | Array(PG::Int16Array) | Array(PG::Int32Array) | Array(PG::Int64Array) | Array(PG::NumericArray) | Array(PG::StringArray) | Array(PG::TimeArray) | Bool | Char | Clear::Expression::UnsafeSql | Float32 | Float64 | Hash(String, JSON::Any) | Int16 | Int32 | Int64 | Int8 | JSON::Any | PG::Geo::Box | PG::Geo::Circle | PG::Geo::Line | PG::Geo::LineSegment | PG::Geo::Path | PG::Geo::Point | PG::Geo::Polygon | PG::Numeric | Slice(UInt8) | String | Time | UInt16 | UInt32 | UInt64 | UInt8 | Nil)#[]=(key : String, value : Time::Span)
```
A trivial fix would be adding `Time::Span` to `Clear::SQL::Any`. But `crystal-pg` doesn't actually support that data type, so this is probably not the best solution.

This needs to be fixed by making sure `clear` only deals with `PG::Connection` instances. I looked at `ConnectionPool` and it's usage and couldn't figure out the entirety of how this is supposed to work. It seems `ConnectionPool` is not really a pool of connections, but a strange pool of `DB::Database` instances. And when it's used in `SQL#transaction`, the transaction itself is always executed on the default database. So I suppose transactions don't work with alternative databaseses at all.

**Comment by anykeyh** [comment]
Please ignore the reference above, I'm referenced a wrong Issue ID.

Clear is made for pgsql only.  I've no will to make it compatible with MySQL in the future.

Concerning the ConnectionPool behavior, it initialize a certain amount of `DB::Database` object and ensure than each fiber perform it works over a different `DB:Database` object.

About `SQL#transaction` there's a optional parameter you can pass to select on which connection you want to start the transaction:

```crystal
SQL.transaction("other_connection") do |cnx|
end
```

It might have a bug however, I must confess the multi-connection system is barely tested.

Regards.

**Comment by straight-shoota** [comment]
> Clear is made for pgsql only. I have no will to make it compatible with MySQL in the future.

Yes I understand that. Clear is obviously targeting psql specifically. I don't want to use clear *with* MySQL, but psql over clear *and* MySQL in the same code base. That should technically be no problem. Both are based on the same crystal-db foundation, but have different implementations.
In order to be a good citizen, clear needs to use the psql implementation `PG::Connection` explicitly instead of the generic `DB:: Connection`.
It seems you don't use connect explicitly and instead use the generic `DB::Databsase` API which implicitly uses `DB:: Connection`. 
AFAIK there's a way to use the `DB::Database` type safe like that. But that API does not really fit to the use case of clear anyways. It adds a lot of overhead because a connection is checked out for every query.
I suppose `ConnectionPool.with_connection` should really yield a connection, i.e. `PG:: Connection` instance. That reduces synchronization overhead, allows o use pg-specific features and doesn't blow up when also using other db drivers.

**Comment by anykeyh** [comment]
You're right. I don't remember why I had to reimplement a connection pooling system, but crystal db comes straight out of the box with it. 
I'll see if I can simplify and use the connection pool of crystal-db instead. It doesn't seem to be a big work

**Comment by straight-shoota** [comment]
`ConnectionPool` isn't even a real connection pool right now. It just disambiguates database configurations (that's what `DB::Database` represents).

**Comment by anykeyh** [comment]
This has been fixed in v0.9. 

Clear is still using its own manager to allocate one connection per fiber, and using `transaction` method from the connection object is not advisable as of now (use of `Clear::SQL.transaction` is required instead).
